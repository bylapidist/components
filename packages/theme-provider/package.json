{
    "name": "@lapidist/theme-provider",
    "version": "0.0.0",
    "description": "ThemeProvider component",
    "author": "Brett Dorrans <hello@lapidist.net>",
    "license": "MIT",
    "main": "src/index.js",
    "scripts": {
        "build": "tsc -p tsconfig.json",
        "test": "jest"
    },
    "publishConfig": {
        "access": "public"
    }
}
