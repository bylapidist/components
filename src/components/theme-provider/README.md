# @lapidist/theme-provider

![npm](https://img.shields.io/npm/v/@lapidist/theme-provider)
