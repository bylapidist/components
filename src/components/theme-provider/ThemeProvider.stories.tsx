import * as React from 'react';
import { ThemeProvider } from '../index';
import { Text } from '../text';

export default {
    title: 'Components/ThemeProvider',
    component: ThemeProvider
};

const templateStyles = {
    padding: '4',
    backgroundColor: { group: 'base', shade: 'light' }
};

const DefaultTemplate = (args) => (
    <ThemeProvider {...args}>
        <Text styles={templateStyles}>
            Nunc porttitor lectus ex, eu pharetra elit placerat non. Suspendisse
            nec ultrices augue, et varius velit. Vestibulum ante ipsum primis in
            faucibus orci luctus et ultrices posuere cubilia curae; Morbi
            sollicitudin sed turpis et vulputate. Nam consequat porttitor
            scelerisque. Nulla ultricies enim at eros accumsan interdum. Aenean
            egestas enim mi, nec tincidunt libero gravida et.
        </Text>
    </ThemeProvider>
);

export const Default = DefaultTemplate.bind({});
Default.args = {};
