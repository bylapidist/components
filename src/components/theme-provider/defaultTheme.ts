export interface ColorGroup {
    lightest: string;
    light: string;
    base: string;
    dark: string;
}

export interface SizingGroup {
    xxxs: string;
    xxs: string;
    xs: string;
    s: string;
    m: string;
    l: string;
    xl: string;
    xxl: string;
    xxxl: string;
}

export interface BreakpointGroup {
    s: string;
    m: string;
    l: string;
}

export interface DefaultTheme {
    colors: {
        greys: ColorGroup;
        blues: ColorGroup;
        yellows: ColorGroup;
        greens: ColorGroup;
        reds: ColorGroup;
    };
    sizing: SizingGroup;
    breakpoints: {
        mobile: BreakpointGroup;
        tablet: BreakpointGroup;
        laptop: BreakpointGroup;
    };
}

export const greys: ColorGroup = {
    lightest: '#ffffff',
    light: '#f3f5f9',
    base: '#66687d',
    dark: '#292b3e'
};

export const blues: ColorGroup = {
    lightest: '#9bfff8',
    light: '#4AD6D8',
    base: '#17a3a5',
    dark: '#007072'
};

export const yellows: ColorGroup = {
    lightest: '#fdffb5',
    light: '#fff08f',
    base: '#ebc764',
    dark: '#c99944'
};

export const greens: ColorGroup = {
    lightest: '#e2f7c9',
    light: '#A7D981',
    base: '#72af5d',
    dark: '#5A8C34'
};

export const reds: ColorGroup = {
    lightest: '#FF848B',
    light: '#E75158',
    base: '#b41e25',
    dark: '#810000'
};

export const defaultTheme: DefaultTheme = {
    colors: {
        greys,
        blues,
        yellows,
        greens,
        reds
    },
    sizing: {
        xxxs: '2px',
        xxs: '4px',
        xs: '10px',
        s: '14px',
        m: '16px',
        l: '24px',
        xl: '36px',
        xxl: '45px',
        xxxl: '57px'
    },
    breakpoints: {
        mobile: {
            s: '320px',
            m: '375px',
            l: '425px'
        },
        tablet: {
            s: '580px',
            m: '640px',
            l: '768px'
        },
        laptop: {
            s: '1024px',
            m: '1280px',
            l: '1440px'
        }
    }
};

export default defaultTheme;
