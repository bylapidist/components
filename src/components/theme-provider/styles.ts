import * as CSS from 'csstype';

import {
    Color,
    getBorderRadius,
    getBorderWidth,
    getBoxShadow,
    getBreakpoint,
    getColor,
    getFontFamily,
    getFontSize,
    getFontWeight,
    getLetterSpacing,
    getLineHeight,
    getSize,
    getWidth,
    NumberOrString,
    Theme,
    ThemeObject
} from './theme';

export interface Styles {
    readonly fontSize?: CSS.FontSizeProperty<NumberOrString>;
    readonly fontFamily?: CSS.FontFamilyProperty;
    readonly fontWeight?: CSS.FontWeightProperty;
    readonly lineHeight?: CSS.LineHeightProperty<NumberOrString>;
    readonly letterSpacing?: CSS.LetterSpacingProperty<NumberOrString>;
    readonly borderRadius?: CSS.BorderRadiusProperty<NumberOrString>;
    readonly borderWidth?: CSS.BorderWidthProperty<NumberOrString>;
    readonly borderTopWidth?: CSS.BorderTopWidthProperty<NumberOrString>;
    readonly borderRightWidth?: CSS.BorderRightWidthProperty<NumberOrString>;
    readonly borderBottomWidth?: CSS.BorderBottomWidthProperty<NumberOrString>;
    readonly borderLeftWidth?: CSS.BorderLeftWidthProperty<NumberOrString>;
    readonly borderColor?: Color;
    readonly borderTopColor?: Color;
    readonly borderRightColor?: Color;
    readonly borderBottomColor?: Color;
    readonly borderLeftColor?: Color;
    readonly backgroundColor?: Color;
    readonly textColor?: Color;
    readonly boxShadow?: CSS.BoxShadowProperty;
    readonly width?: CSS.WidthProperty<NumberOrString>;
    readonly minWidth?: CSS.WidthProperty<NumberOrString>;
    readonly maxWidth?: CSS.WidthProperty<NumberOrString>;
    readonly textAlign?: CSS.TextAlignProperty;
    readonly margin?: CSS.MarginProperty<NumberOrString>;
    readonly marginX?: CSS.MarginProperty<NumberOrString>;
    readonly marginY?: CSS.MarginProperty<NumberOrString>;
    readonly marginTop?: CSS.MarginTopProperty<NumberOrString>;
    readonly marginRight?: CSS.MarginRightProperty<NumberOrString>;
    readonly marginBottom?: CSS.MarginBottomProperty<NumberOrString>;
    readonly marginLeft?: CSS.MarginLeftProperty<NumberOrString>;
    readonly padding?: CSS.PaddingProperty<NumberOrString>;
    readonly paddingX?: CSS.PaddingProperty<NumberOrString>;
    readonly paddingY?: CSS.PaddingProperty<NumberOrString>;
    readonly paddingTop?: CSS.PaddingTopProperty<NumberOrString>;
    readonly paddingRight?: CSS.PaddingRightProperty<NumberOrString>;
    readonly paddingBottom?: CSS.PaddingBottomProperty<NumberOrString>;
    readonly paddingLeft?: CSS.PaddingLeftProperty<NumberOrString>;
    readonly breakpoints?: ThemeObject<Styles>;
}

const withTypography = (theme: Theme, styles?: Styles): string =>
    styles
        ? `
    ${
        styles.fontSize
            ? `font-size: ${getFontSize(theme, styles.fontSize)};`
            : ''
    }

        ${
            styles.fontWeight
                ? `font-weight: ${getFontWeight(theme, styles.fontWeight)};`
                : ''
        }

        ${
            styles.fontFamily
                ? `font-family: ${getFontFamily(theme, styles.fontFamily)};`
                : ''
        }

        ${
            styles.lineHeight
                ? `line-height: ${getLineHeight(theme, styles.lineHeight)};`
                : ''
        }

        ${
            styles.letterSpacing
                ? `letter-spacing: ${getLetterSpacing(
                      theme,
                      styles.letterSpacing
                  )};`
                : ''
        }

        ${
            styles.textColor
                ? `text-color: ${getColor(
                      theme,
                      styles.textColor.colorGroup,
                      styles.textColor.colorShade
                  )};`
                : ''
        }
    `
        : '';

const withBorders = (theme: Theme, styles?: Styles): string =>
    styles
        ? `
       ${
           styles.borderRadius
               ? `border-radius: ${getBorderRadius(
                     theme,
                     styles.borderRadius
                 )};`
               : ''
       }

        ${
            styles.borderWidth
                ? `border-width: ${getBorderWidth(theme, styles.borderWidth)};`
                : ''
        }

        ${
            styles.borderTopWidth
                ? `border-top-width: ${getBorderWidth(
                      theme,
                      styles.borderTopWidth
                  )};`
                : ''
        }

        ${
            styles.borderRightWidth
                ? `border-right-width: ${getBorderWidth(
                      theme,
                      styles.borderRightWidth
                  )};`
                : ''
        }

        ${
            styles.borderBottomWidth
                ? `border-bottom-width: ${getBorderWidth(
                      theme,
                      styles.borderBottomWidth
                  )};`
                : ''
        }

        ${
            styles.borderLeftWidth
                ? `border-left-width: ${getBorderWidth(
                      theme,
                      styles.borderLeftWidth
                  )};`
                : ''
        }

        ${
            styles.boxShadow
                ? `box-shadow: ${getBoxShadow(theme, styles.boxShadow)};`
                : ''
        }
    `
        : '';

const withLayout = (theme: Theme, styles?: Styles): string =>
    styles
        ? `
        ${styles.width ? `width: ${getWidth(theme, styles.width)};` : ''}

        ${
            styles.minWidth
                ? `min-width: ${getWidth(theme, styles.minWidth)};`
                : ''
        }

        ${
            styles.maxWidth
                ? `max-width: ${getWidth(theme, styles.maxWidth)};`
                : ''
        }

        ${styles.textAlign ? `text-align: ${styles.textAlign};` : ''}

        ${styles.margin ? `margin: ${getSize(theme, styles.margin)};` : ''}

        ${
            styles.marginX
                ? `margin-left: ${getSize(theme, styles.marginX)};`
                : ''
        }

        ${
            styles.marginX
                ? `margin-right: ${getSize(theme, styles.marginX)};`
                : ''
        }

        ${
            styles.marginY
                ? `margin-top: ${getSize(theme, styles.marginY)};`
                : ''
        }

        ${
            styles.marginY
                ? `margin-bottom: ${getSize(theme, styles.marginY)};`
                : ''
        }

        ${
            styles.marginTop
                ? `margin-top: ${getSize(theme, styles.marginTop)};`
                : ''
        }

        ${
            styles.marginRight
                ? `margin-right: ${getSize(theme, styles.marginRight)};`
                : ''
        }

        ${
            styles.marginBottom
                ? `margin-bottom: ${getSize(theme, styles.marginBottom)};`
                : ''
        }

        ${
            styles.marginLeft
                ? `margin-left: ${getSize(theme, styles.marginLeft)};`
                : ''
        }

        ${styles.padding ? `padding: ${getSize(theme, styles.padding)};` : ''}

        ${
            styles.paddingX
                ? `padding-left: ${getSize(theme, styles.paddingX)};`
                : ''
        }

        ${
            styles.paddingX
                ? `padding-right: ${getSize(theme, styles.paddingX)};`
                : ''
        }

        ${
            styles.paddingY
                ? `padding-top: ${getSize(theme, styles.paddingY)};`
                : ''
        }

        ${
            styles.paddingY
                ? `padding-bottom: ${getSize(theme, styles.paddingY)};`
                : ''
        }

        ${
            styles.paddingTop
                ? `padding-top: ${getSize(theme, styles.paddingTop)};`
                : ''
        }

        ${
            styles.paddingRight
                ? `padding-right: ${getSize(theme, styles.paddingRight)};`
                : ''
        }

        ${
            styles.paddingBottom
                ? `padding-bottom: ${getSize(theme, styles.paddingBottom)};`
                : ''
        }

        ${
            styles.paddingLeft
                ? `padding-left: ${getSize(theme, styles.paddingLeft)};`
                : ''
        }
    `
        : '';

const withColors = (theme: Theme, styles?: Styles): string =>
    styles
        ? `
        ${
            styles.backgroundColor
                ? `background-color: ${getColor(
                      theme,
                      styles.backgroundColor.colorGroup,
                      styles.backgroundColor.colorShade
                  )};`
                : ''
        }

        ${
            styles.borderTopColor
                ? `border-top-color: ${getColor(
                      theme,
                      styles.borderTopColor.colorGroup,
                      styles.borderTopColor.colorShade
                  )};`
                : ''
        }

        ${
            styles.borderRightColor
                ? `border-right-color: ${getColor(
                      theme,
                      styles.borderRightColor.colorGroup,
                      styles.borderRightColor.colorShade
                  )};`
                : ''
        }

        ${
            styles.borderBottomColor
                ? `border-bottom-color: ${getColor(
                      theme,
                      styles.borderBottomColor.colorGroup,
                      styles.borderBottomColor.colorShade
                  )};`
                : ''
        }

        ${
            styles.borderLeftColor
                ? `border-left-color: ${getColor(
                      theme,
                      styles.borderLeftColor.colorGroup,
                      styles.borderLeftColor.colorShade
                  )};`
                : ''
        }

        ${
            styles.borderColor
                ? `border-color: ${getColor(
                      theme,
                      styles.borderColor.colorGroup,
                      styles.borderColor.colorShade
                  )};`
                : ''
        }
    `
        : '';

export const withBaseStyles = (theme: Theme, styles?: Styles): string =>
    styles
        ? `
        ${withTypography(theme, styles)}
        ${withBorders(theme, styles)}
        ${withLayout(theme, styles)}
        ${withColors(theme, styles)}
`
        : '';

const withMediaQueries = (
    breakpoints: ThemeObject<Styles>,
    theme: Theme
): string =>
    Object.entries(breakpoints)
        .map(
            ([breakpoint, styles]) => `
            @media only screen and (min-width: ${getBreakpoint(
                theme,
                breakpoint
            )}) {
                ${withBaseStyles(theme, styles)}
            }
        `
        )
        .join('');

const withBreakpoints = (theme: Theme, styles?: Styles): string =>
    styles && styles.breakpoints
        ? `
        ${withMediaQueries(styles.breakpoints, theme)}
    `
        : '';

export const withStyles = (theme: Theme, styles?: Styles): string =>
    styles
        ? `
        ${withBaseStyles(theme, styles)}
        ${withBreakpoints(theme, styles)}
`
        : '';
