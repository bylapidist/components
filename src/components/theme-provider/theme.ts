import * as CSS from 'csstype';
import { defaultTheme } from './defaultTheme';

type ThemeObject<T> = { [K: string]: T };

type ThemeArray<T> = { [K: string]: T[] };

export type NumberOrString = string | number | 0;

export type Color = { colorGroup: NumberOrString; colorShade: NumberOrString };

export interface Theme {
    breakpoints?: ThemeObject<string>;
    fontSizes?: ThemeObject<CSS.FontSizeProperty<NumberOrString>>;
    fontFamilies?: ThemeArray<CSS.FontFamilyProperty>;
    fontWeights?: ThemeObject<CSS.FontWeightProperty>;
    lineHeights?: ThemeObject<CSS.LineHeightProperty<NumberOrString>>;
    letterSpacings?: ThemeObject<CSS.LetterSpacingProperty<NumberOrString>>;
    sizes?: ThemeObject<string>;
    borderRadii?: ThemeObject<CSS.BorderRadiusProperty<NumberOrString>>;
    borderWidths?: ThemeObject<CSS.BorderWidthProperty<NumberOrString>>;
    boxShadows?: ThemeObject<CSS.BoxShadowProperty>;
    widths?: ThemeObject<string>;
    colors?: ThemeObject<ThemeObject<CSS.ColorProperty>>;
}

export const mergeThemes = (theme?: Theme): Theme => {
    if (!theme) return defaultTheme;

    return { ...theme, ...defaultTheme };
};

export const getBreakpoint = (
    theme: Theme,
    breakpoint: NumberOrString
): string => {
    if (theme.breakpoints && theme.breakpoints[breakpoint]) {
        return theme.breakpoints[breakpoint];
    }

    throw new Error(`${breakpoint} does not exist on theme.breakpoints`);
};

export const getAllBreakpoints = (theme: Theme): string[] => {
    if (theme.breakpoints) {
        return Object.keys(theme.breakpoints);
    }

    throw new Error(`theme.breakpoints does not exist`);
};

export const getAllMediaQueries = (theme: Theme): ThemeObject<string> => {
    if (theme.breakpoints) {
        const allBreakpoints = getAllBreakpoints(theme);
        const allMediaQueries: ThemeObject<string> = {};

        allBreakpoints.forEach((breakpoint) => {
            allMediaQueries[breakpoint] = `(min-width: ${getBreakpoint(
                theme,
                breakpoint
            )})`;
        });

        return allMediaQueries;
    }

    throw new Error(`theme.breakpoints does not exist`);
};

export const getFontSize = (
    theme: Theme,
    size: NumberOrString
): CSS.FontSizeProperty<NumberOrString> => {
    if (theme.fontSizes && theme.fontSizes[size]) {
        return theme.fontSizes[size];
    }

    throw new Error(`${size} does not exist on theme.fontSizes`);
};

export const getFontFamily = (
    theme: Theme,
    fontFamily: NumberOrString
): CSS.FontFamilyProperty => {
    if (theme.fontFamilies && theme.fontFamilies[fontFamily]) {
        return theme.fontFamilies[fontFamily].join(', ');
    }

    throw new Error(`${fontFamily} does not exist on theme.fontFamilies`);
};

export const getFontWeight = (
    theme: Theme,
    fontWeight: NumberOrString
): CSS.FontWeightProperty => {
    if (theme.fontWeights && theme.fontWeights[fontWeight]) {
        return theme.fontWeights[fontWeight];
    }

    throw new Error(`${fontWeight} does not exist on theme.fontWeights`);
};

export const getLineHeight = (
    theme: Theme,
    lineHeight: NumberOrString
): CSS.LineHeightProperty<NumberOrString> => {
    if (theme.lineHeights && theme.lineHeights[lineHeight]) {
        return theme.lineHeights[lineHeight];
    }

    throw new Error(`${lineHeight} does not exist on theme.lineHeights`);
};

export const getLetterSpacing = (
    theme: Theme,
    letterSpacing: NumberOrString
): CSS.LetterSpacingProperty<NumberOrString> => {
    if (theme.letterSpacings && theme.letterSpacings[letterSpacing]) {
        return theme.letterSpacings[letterSpacing];
    }

    throw new Error(`${letterSpacing} does not exist on theme.letterSpacings`);
};

export const getSize = (theme: Theme, size: NumberOrString): string => {
    if (theme.sizes && theme.sizes[size]) {
        return theme.sizes[size];
    }

    throw new Error(`${size} does not exist on theme.sizes`);
};

export const getBorderRadius = (
    theme: Theme,
    borderRadius: NumberOrString
): CSS.BorderRadiusProperty<NumberOrString> => {
    if (theme.borderRadii && theme.borderRadii[borderRadius]) {
        return theme.borderRadii[borderRadius];
    }

    throw new Error(`${borderRadius} does not exist on theme.borderRadii`);
};

export const getBorderWidth = (
    theme: Theme,
    borderWidth: NumberOrString
): CSS.BorderWidthProperty<NumberOrString> => {
    if (theme.borderWidths && theme.borderWidths[borderWidth]) {
        return theme.borderWidths[borderWidth];
    }

    throw new Error(`${borderWidth} does not exist on theme.borderWidths`);
};

export const getBoxShadow = (
    theme: Theme,
    boxShadow: NumberOrString
): CSS.BoxShadowProperty => {
    if (theme.boxShadows && theme.boxShadows[boxShadow]) {
        return theme.boxShadows[boxShadow];
    }

    throw new Error(`${boxShadow} does not exist on theme.boxShadows`);
};

export const getWidth = (
    theme: Theme,
    width: NumberOrString
): CSS.WidthProperty<NumberOrString> => {
    if (theme.widths && theme.widths[width]) {
        return theme.widths[width];
    }

    throw new Error(`${width} does not exist on theme.widths`);
};

export const getColor = (
    theme: Theme,
    colorGroup: NumberOrString,
    colorShade: NumberOrString
): CSS.ColorProperty => {
    if (
        theme.colors &&
        theme.colors[colorGroup] &&
        theme.colors[colorGroup][colorShade]
    ) {
        return theme.colors[colorGroup][colorShade];
    }

    throw new Error(
        `${colorGroup} or ${colorShade} does not exist on theme.colors`
    );
};
