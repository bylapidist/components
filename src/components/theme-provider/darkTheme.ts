import { Theme } from '@lapidist/styles';

export const darkTheme: Theme = {
    boxShadows: {
        '1':
            '0 1px 3px 0 rgba(255, 255, 255, 0.1), 0 1px 2px 0 rgba(255, 255, 255, 0.06)'
    },
    colors: {
        base: {
            light: 'black',
            dark: 'white',
            transparent: 'transparent'
        },
        grey: {
            lightest: '#292B3E',
            light: '#73748b',
            base: '#d4d4d4',
            dark: '#E8E8E8'
        },
        primary: {
            lightest: '#1a8385',
            light: '#17a3a5',
            base: '#4AD6D8',
            dark: '#9bfff8'
        },
        secondary: {
            lightest: '#c99944',
            light: '#ebc764',
            base: '#fff08f',
            dark: '#fdffb5'
        },
        tertiary: {
            lightest: '#5A8C34',
            light: '#72af5d',
            base: '#A7D981',
            dark: '#e2f7c9'
        },
        danger: {
            lightest: '#810000',
            light: '#b41e25',
            base: '#E75158',
            dark: '#FF848B'
        }
    }
};
