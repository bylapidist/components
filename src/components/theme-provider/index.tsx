import React, { Context, PropsWithChildren } from 'react';
import {
    ThemeProvider as StyledThemeProvider,
    createGlobalStyle
} from 'styled-components';
import { mergeThemes, Theme } from '@lapidist/styles';
import { defaultTheme } from './defaultTheme';
import { darkTheme } from './darkTheme';

const GlobalStyle = createGlobalStyle`
    html, body, div, span, applet, object, iframe,
    h1, h2, h3, h4, h5, h6, p, blockquote, pre,
    a, abbr, acronym, address, big, cite, code,
    del, dfn, em, img, ins, kbd, q, s, samp,
    small, strike, strong, sub, sup, tt, var,
    b, u, i, center, button,
    dl, dt, dd, menu, ol, ul, li,
    fieldset, form, label, legend,
    table, caption, tbody, tfoot, thead, tr, th, td,
    article, aside, canvas, details, embed,
    figure, figcaption, footer, header, hgroup,
    main, menu, nav, output, ruby, section, summary,
    time, mark, audio, video {
        margin: 0;
        padding: 0;
        border: 0;
        font-size: 100%;
        font: inherit;
        vertical-align: baseline;
        transition: color 0.12s cubic-bezier(0.6, 0.04, 0.98, 0.7) 0s,
            border 0.12s cubic-bezier(0.6, 0.04, 0.98, 0.7) 0s,
            background-color 0.12s cubic-bezier(0.6, 0.04, 0.98, 0.7) 0s;
    }
    html { box-sizing: border-box; }
    *, *:before, *:after { box-sizing: inherit; }
    article, aside, details, figcaption, figure,
    footer, header, hgroup, main, menu, nav, section { display: block; }
    *[hidden] { display: none; }
    body { line-height: 1; font-size: 62.5%; }
    menu, ol, ul { list-style: none; }
    blockquote, q { quotes: none; }
    blockquote:before, blockquote:after,
    q:before, q:after { content: ''; content: none; }
    table { border-collapse: collapse; border-spacing: 0; }
    img { display: block; max-width: 100%; }
    a, button { text-decoration: none; outline-offset: 2px; outline: transparent solid 2px; }
    a:focus, button:focus { outline: rgb(235, 199, 100) solid 2px; }
`;

export interface ThemeProviderProps extends PropsWithChildren<unknown> {
    readonly theme?: Theme;
}

export const ThemeContext: Context<Theme> =
    React.createContext<Theme>(defaultTheme);

export const useTheme = () => React.useContext(ThemeContext);

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
    children,
    theme
}) => {
    const initialTheme: Theme = useTheme();
    const baseTheme: Theme | undefined = mergeThemes(initialTheme, theme);

    const [isDarkMode, setIsDarkMode] = React.useState<boolean>();
    const [themeContext, setThemeContext] =
        React.useState<Theme | undefined>(baseTheme);

    React.useEffect(() => {
        const isDarkModePersisted =
            localStorage.getItem('isDarkMode') === 'true';
        setIsDarkMode(isDarkModePersisted);
    }, [setIsDarkMode]);

    React.useEffect(() => {
        setThemeContext(
            isDarkMode ? mergeThemes(baseTheme, darkTheme) : baseTheme
        );
    }, [isDarkMode, baseTheme]);

    return (
        <>
            <GlobalStyle />
            <StyledThemeProvider theme={themeContext}>
                {children}
            </StyledThemeProvider>
        </>
    );
};

ThemeProvider.displayName = 'ThemeProvider';
