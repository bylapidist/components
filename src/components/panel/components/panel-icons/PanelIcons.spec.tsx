import * as React from 'react';
import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';

import { ThemeProvider } from '../../../theme-provider';
import { PanelIcons } from './index';
import { PanelStatusType } from '../../index';

const setup = (Component: React.ReactElement) =>
    render(<ThemeProvider>{Component}</ThemeProvider>);

afterEach(cleanup);

const statuses: PanelStatusType[] = ['none', 'info', 'warning', 'error'];

const defaultProps = {
    status: undefined,
    tag: undefined
};

test('it works', () => {
    const { container } = setup(<PanelIcons {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
});

test.each(statuses)('it works with status', (status) => {
    const { container } = setup(<PanelIcons status={status} />);
    expect(container.firstChild).toMatchSnapshot();
});

test('it works with tag', () => {
    const { container, getByText } = setup(
        <PanelIcons
            {...defaultProps}
            tag={{ title: 'v1.0.0', props: { kind: 'primary' } }}
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(getByText('v1.0.0')).toBeTruthy();
});
