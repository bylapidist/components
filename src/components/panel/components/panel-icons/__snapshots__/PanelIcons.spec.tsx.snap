// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it works 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
/>
`;

exports[`it works with status 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
/>
`;

exports[`it works with status 2`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 1rem;
  font-family: Montserrat,sans-serif;
  line-height: 2;
  font-weight: 400;
  color: #292B3E;
}

.c1 > p {
  margin-bottom: 0.5rem;
}

.c1 > p:last-child {
  margin-bottom: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-info "
      data-icon="circle-info"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 0C114.6 0 0 114.6 0 256s114.6 256 256 256s256-114.6 256-256S397.4 0 256 0zM256 128c17.67 0 32 14.33 32 32c0 17.67-14.33 32-32 32S224 177.7 224 160C224 142.3 238.3 128 256 128zM296 384h-80C202.8 384 192 373.3 192 360s10.75-24 24-24h16v-64H224c-13.25 0-24-10.75-24-24S210.8 224 224 224h32c13.25 0 24 10.75 24 24v88h16c13.25 0 24 10.75 24 24S309.3 384 296 384z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`it works with status 3`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 1rem;
  font-family: Montserrat,sans-serif;
  line-height: 2;
  font-weight: 400;
  color: #292B3E;
}

.c1 > p {
  margin-bottom: 0.5rem;
}

.c1 > p:last-child {
  margin-bottom: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-exclamation "
      data-icon="circle-exclamation"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 0C114.6 0 0 114.6 0 256s114.6 256 256 256s256-114.6 256-256S397.4 0 256 0zM232 152C232 138.8 242.8 128 256 128s24 10.75 24 24v128c0 13.25-10.75 24-24 24S232 293.3 232 280V152zM256 400c-17.36 0-31.44-14.08-31.44-31.44c0-17.36 14.07-31.44 31.44-31.44s31.44 14.08 31.44 31.44C287.4 385.9 273.4 400 256 400z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`it works with status 4`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 1rem;
  font-family: Montserrat,sans-serif;
  line-height: 2;
  font-weight: 400;
  color: #292B3E;
}

.c1 > p {
  margin-bottom: 0.5rem;
}

.c1 > p:last-child {
  margin-bottom: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-exclamation "
      data-icon="circle-exclamation"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 0C114.6 0 0 114.6 0 256s114.6 256 256 256s256-114.6 256-256S397.4 0 256 0zM232 152C232 138.8 242.8 128 256 128s24 10.75 24 24v128c0 13.25-10.75 24-24 24S232 293.3 232 280V152zM256 400c-17.36 0-31.44-14.08-31.44-31.44c0-17.36 14.07-31.44 31.44-31.44s31.44 14.08 31.44 31.44C287.4 385.9 273.4 400 256 400z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`it works with tag 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.c2 {
  font-size: 1rem;
  font-family: Montserrat,sans-serif;
  line-height: 2;
  font-weight: 400;
  color: #292B3E;
}

.c2 > p {
  margin-bottom: 0.5rem;
}

.c2 > p:last-child {
  margin-bottom: 0;
}

.c3 {
  text-align: center;
  border-style: solid;
  border-width: 0;
  padding: 0.25rem 0;
  color: white;
  font-size: 0.75rem;
  font-weight: 500;
  border-radius: 0.6rem;
  background-color: #135f61;
}

.c4 {
  padding: 0.25rem 0.5rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2 c3"
      kind="primary"
    >
      <span
        class="c4"
        kind="primary"
      >
        v1.0.0
      </span>
    </span>
  </div>
</div>
`;
